components/eamxx/src/diagnostics/field_at_height.cpp

```diff
@@ -0,0 +1,164 @@
+
+  // Note: you may ask why we can't just store f and be done, rather than go through the
+  // add_field infrastructure. Unfortunately, there are some checks in the base classes
+  // that require the diagnostic to have 1+ required fields. So we have to do this.
+  // TODO: one day we may make atm diags *not* inherit from atm process...
+  add_field<Required>(fid);
+
+  // We can also create the diagnostic already!
+  const auto& location = params.get<std::string>("Field Level Location");
+  const auto diag_field_name = m_field_name + "_at_" + location;
+
+  FieldIdentifier d_fid (diag_field_name,layout.strip_dim(tag),fid.get_units(),fid.get_grid_name());
+  m_diagnostic_output = Field(d_fid);
+  m_diagnostic_output.allocate_view();
+
+  // Figure out the pressure value
+  auto chars_start = location.find_first_not_of("0123456789.");
+  EKAT_REQUIRE_MSG (chars_start!=0 && chars_start!=std::string::npos,
+      "Error! Invalid string for pressure value for FieldAtHeight.\n"
+      " - input string   : " + location + "\n"
+      " - expected format: Nm, with N integer\n");
+  const auto z_str = location.substr(0,chars_start);
+  m_z = std::stod(z_str);
+
+  const auto units = location.substr(chars_start);
+  EKAT_REQUIRE_MSG (units=="m",
+      "Error! Invalid string for height value for FieldAtHeight.\n"
+      " - input string   : " + location + "\n"
+      " - expected format: Nm, with N integer\n");
+
+  // Create request for z field
+  const auto& gname = fid.get_grid_name();
+  m_z_name = tag==LEV ? "z_mid" : "z_int";
+  add_field<Required>(m_z_name, layout, ekat::units::m, gname);
+}
+
+// =========================================================================================
+void FieldAtHeight::compute_diagnostic_impl()
+{
+  const auto z_view = get_field_in(m_z_name).get_view<const Real**>();
+  const Field& f = get_field_in(m_field_name);
+  const auto& fl = f.get_header().get_identifier().get_layout();
+
+  using RangePolicy = typename KokkosTypes<DefaultDevice>::RangePolicy;
+
+  auto z_tgt = m_z;
+  auto nlevs = fl.dims().back();
+  if (fl.rank()==2) {
+    const auto f_view = f.get_view<const Real**>();
+    const auto d_view = m_diagnostic_output.get_view<Real*>();
+
+    RangePolicy policy (0,fl.dims()[0]);
+    Kokkos::parallel_for(policy,
+        KOKKOS_LAMBDA(const int i) {
+        auto f_i = ekat::subview(f_view,i);
+        auto z_i = ekat::subview(z_view,i);
+
+        auto beg = z_i.data();
+        auto end = beg+nlevs;
+        auto it = find_first_smaller_z(beg,end,z_tgt);
+        if (it==beg) {
+          // We just extapolate *beg
+          d_view(i) = *beg;
+        } else if (it==end) {
+          // We just extapolate *end
+          d_view(i) = *(--end);
+        } else {
+          auto pos = it-beg;
+          auto z0 = z_i(pos-1);
+          auto z1 = z_i(pos);
+          auto f0 = f_i(pos-1);
+          auto f1 = f_i(pos);
+
+          d_view(i) = ( (z_tgt-z0)*f1 + (z1-z_tgt)*f0 ) / (z1-z0);
```

Comment:
Is there a reason not to use EKAT's linear interpolation utility here? That would provide the benefit of a unit-tested routine.

Reply:
I thought about it, but given that there is one grid point, I thought the hierarchical ||ism was a bit of a waste, and since the data structures won't be reused, there was no advantage in reusing internal data computed by LinInterp.